onadata/apps/api/viewsets/data_viewset.py

```diff
@@ -633,87 +707,80 @@ def list(self, request, *args, **kwargs):
 
         return custom_response_handler(request, xform, query, export_type)
 
-    def modify(self, request, *args, **kwargs):
+    @staticmethod
+    def __get_payload(request):
+        try:
+            payload = json.loads(request.data.get('payload', '{}'))
+        except ValueError:
+            raise ParseError(_('Invalid payload'))
 
-        xform = self.get_object()
-        http_status = status.HTTP_200_OK
-        response = {}
+        return payload
+
+    @staticmethod
+    def __get_instances_ids(payload, mongo_base_query):
+
+        """
+        Gets instances ids based on the request payload.
+        Useful to narrow down set of instances for bulk actions
+
+        Args:
+            payload (dict)
+            mongo_base_query (dict): Base query which must contains '_userform_id'
+
+        Returns:
+            list|None: `None` means all instances are returned. Empty list means
+                        no results found
+
+        """
 
-        if request.user.has_perm("validate_xform", xform):
+        instances_ids = None
 
-            owner = xform.user
-            userform_id = "{}_{}".format(owner.username, xform.id_string)
-            query = {ParsedInstance.USERFORM_ID: userform_id}  # Query used for MongoDB
-            filter_ = {"xform_id": xform.id}  # Filter for Django ORM
-            payload = {}
+        ###################################################
+        # Submissions can be retrieve in 3 different ways #
+        ###################################################
 
+        # First scenario / Get submissions based on user's query
+        if payload.get('query'):
+            # Validate if query is valid.
             try:
-                payload = json.loads(request.data.get("payload", "{}"))
+                query = payload.get('query')
```

Comment:
Instead of the `if payload.get('query')…query = payload.get('query')` construct, why not:
```python
try:
    query = payload['query']
except KeyError:
    pass # No query? No problem!
else:
    # Make sure query is a dict, update it with the base query, run it in Mongo, etc.
    …

Reply:
[EAFP](https://docs.python.org/3.7/glossary.html#term-eafp) is not natural for me.  Because it seems to be the pythonic way, I'll do my best to integrate it in my code ;-) 
Interesting discussion in [stackoverflow - Using try vs if in python](https://stackoverflow.com/a/1835844)

Reply:
You argue my point better than I do :wink: Thanks for the glossary link to EAFP vs LBYL.

I did learn something new about the cost of handling (or raising?) an exception, but I'd add that one thing is missing from your benchmark: the second retrieval from the dictionary inside the `if` block. For curiosity's sake, I ran these on my machine (which looks to be slower than yours). First, unmodified except for using `print`, so I could paste the whole block into the interpreter at once:
```
In [2]: import timeit
   ...: print(timeit.timeit(setup="payload={'foo': 'bar'}", stmt="try:\n payload['query']\nexcept KeyError:\n pass"))
   ...: print(timeit.timeit(setup="payload={'foo': 'bar'}", stmt="if payload.get('query'):\n pass"))
   ...: print(timeit.timeit(setup="payload={'query': 'bar'}", stmt="if payload.get('query'):\n pass"))
   ...: print(timeit.timeit(setup="payload={'query': 'bar'}", stmt="try:\n payload['query']\nexcept KeyError:\n pass"))
   ...: 
0.628930091858
0.0932941436768
0.109492063522
0.0401449203491
```
Now, doing assignments in the `try` and `if` blocks:
```
In [3]: import timeit
   ...: print(timeit.timeit(setup="payload={'foo': 'bar'}", stmt="try:\n pl = payload['query']\nexcept KeyError:\n pass"))
   ...: print(timeit.timeit(setup="payload={'foo': 'bar'}", stmt="if payload.get('query'):\n pl = payload['query']"))
   ...: print(timeit.timeit(setup="payload={'query': 'bar'}", stmt="if payload.get('query'):\n pl = payload['query']"))
   ...: print(timeit.timeit(setup="payload={'query': 'bar'}", stmt="try:\n pl = payload['query']\nexcept KeyError:\n pass"))
   ...: 
0.602092981339
0.109379053116
0.152740955353
0.0410978794098
```
No earth-shaking realizations here :slightly_smiling_face: